app/main/views/questionnaire.py

```diff
@@ -22,79 +18,90 @@
-        metadata_survey = metadata.eq_id + metadata.form_type + metadata.period_id + metadata.collection_exercise_sid
+        logger.info("Redirecting user to next location %s with tx_id=%s", next_location, metadata.tx_id)
+        return do_redirect(eq_id, form_type, period_id, collection_id, next_location)
+    except InvalidLocationException:
+        return do_redirect(eq_id, form_type, period_id, collection_id, questionnaire_manager.get_current_location())
 
-        if not current_survey == metadata_survey:
-            return redirect("/information/multiple-surveys")
 
-        # Redirect to thank you page if the questionnaire has already been submitted
-        if questionnaire_manager.submitted and location != 'thank-you':
-            return do_redirect(eq_id, form_type, period_id, collection_id, 'thank-you')
+@main_blueprint.route('/questionnaire/<eq_id>/<form_type>/<period_id>/<collection_id>/previous', methods=['GET'])
+@login_required
+@check_survey_state
+def go_to_previous_page(eq_id, form_type, period_id, collection_id):
+    questionnaire_manager = QuestionnaireManagerFactory.get_instance()
+    previous_location = questionnaire_manager.get_previous_location()
+    questionnaire_manager.go_to(previous_location)
+    return do_redirect(eq_id, form_type, period_id, collection_id, previous_location)
 
-        # redirect to the first block if the change your answers link is clicked
-        if location == 'first' or location == 'previous':
-            location = questionnaire_manager.resolve_location(location)
-            questionnaire_manager.go_to(location)
-            return do_redirect(eq_id, form_type, period_id, collection_id, location)
 
-        # Process the POST request
-        if request.method == 'POST':
-            return do_post(collection_id, eq_id, form_type, period_id, location, questionnaire_manager)
-        else:
-            return do_get(questionnaire_manager, location)
+@main_blueprint.route('/questionnaire/<eq_id>/<form_type>/<period_id>/<collection_id>/thank-you', methods=["GET"])
+@login_required
+def get_thank_you(eq_id, form_type, period_id, collection_id):
+    # Delete user data on request of thank you page.
+    questionnaire_manager = QuestionnaireManagerFactory.get_instance()
+    if not same_survey(eq_id, form_type, period_id, collection_id):
+        return redirect("/information/multiple-surveys")
 
-    except QuestionnaireException as e:
-        return page_not_found(e)
-    except InvalidLocationException as e:
-        return do_redirect(eq_id, form_type, period_id, collection_id, questionnaire_manager.get_current_location())
-    except SubmissionFailedException as e:
-        # Rabbit MQ connection issue
-        return service_unavailable(e)
-    except Exception as e:
-        return internal_server_error(e)
+    page = render_page('thank-you', questionnaire_manager)
+    questionnaire_manager.delete_user_data()
```

Comment:
I feel that it may be good to move the questionnaire manager behaviour behind into this main route? It's only ever called here.